File: lazarus/scripts/systems/particles_system.js
class Particle {
    constructor(){
	this.graphics = null
	this.type = ""
	this.x = 0
	this.y = 0
	this.rot = 0
	this.v = 0
	this.size = 1
	this.lifetime = 0
	this.max_lifetime = 0
	this.next = null
    }
    init(graphics, type, x, y, rot, v, size, lifetime){
	this.graphics = graphics
	this.type = type
	this.x = x
	this.y = y
	this.rot = rot
	this.v = v
	this.size = size
	this.lifetime = lifetime
	this.max_lifetime = lifetime
    }
    in_use(){
	return this.lifetime > 0
    }
    update(dt){
	if(!this.in_use()){
	    return false
	}
	// Movement?
	this.lifetime -= dt
	return this.lifetime <= 0
    }
}

class ParticlesPool {
    constructor(pool_size){
	this.particles = []
	this.pool_size = pool_size
	for(var i = 0; i<pool_size; i++){
	    this.particles.push(new Particle())
	}
	this.first_available = this.particles[0]
	for(var i = 0; i<pool_size-1; i++){
	    this.particles[i].next = this.particles[i+1]
	}
    }
    create(graphics, type, x, y, rot, v, size, lifetime){
	if(this.first_available === null){
	    console.log("particle pool is full")
	    return
	}

	var new_particle = this.first_available
	this.first_available = new_particle.next

	new_particle.init(graphics, type, x, y, rot, v, size, lifetime)
    }
    update(dt){
	for(var particle of this.particles){
	    if(particle.update(dt)){ // particle is dead
		particle.next = this.first_available
		this.first_available = particle
	    }
	}
    }
}
	    
	
export var particles_pool = new ParticlesPool(1000)

File: fiddling/server/systems/particles_system.py
class PartPos:
    def __init__(self, x, y, rot, size_x, size_y):
        self.x = x
        self.y = y
        self.rot = rot
        self.size_x = size_x
        self.size_y = size_y


class Particle:
    def __init__(self):
        self.type = ""
        self.position = PartPos(0, 0, 0, 1, 1)
        self.v = 0
        self.lifetime = 0
        self.max_lifetime = 0
        self.next = None

    def init(self, type, x, y, rot, v, size_x, size_y, lifetime):
        self.type = type
        self.position.x = x
        self.position.y = y
        self.position.rot = rot
        self.v = v
        self.position.size_x = size_x
        self.position.size_y = size_y
        self.lifetime = lifetime
        self.max_lifetime = lifetime

    def in_use(self):
        return self.lifetime > 0

    def update(self, dt):
        if not self.in_use():
            return False

        # Movement??
        self.lifetime -= dt
        return self.lifetime <= 0


class ParticlesPool:
    def __init__(self, pool_size):
        self.particles = []
        self.pool_size = pool_size
        for i in range(pool_size):
            self.particles.append(Particle())
        self.first_available = self.particles[0]
        for i in range(pool_size - 1):
            self.particles[i].next = self.particles[i+1]

    def create(self, type, x, y, rot, v, size_x, size_y, lifetime):
        if self.first_available is None:
            print("Particle pool is full")
            return

        new_particle = self.first_available
        self.first_available = new_particle.next

        new_particle.init(type, x, y, rot, v, size_x, size_y, lifetime)

    def update(self, world, dt):
        for particle in self.particles:
            if particle.update(dt):
                # particle is dead
                particle.next = self.first_available
                self.first_available = particle

    def _serialise(self):
        ret = []
        # TODO: improve this so that we don't loop on all particles
        for particle in self.particles:
            if particle.in_use():
                ret.append({
                    "type": particle.type,
                    "x": particle.position.x,
                    "y": particle.position.y,
                    "rot": particle.position.rot,
                    "size_x": particle.position.size_x,
                    "size_y": particle.position.size_y
                    })
        return ret


particles_pool = ParticlesPool(1000)
